### PR TITLE
Add ERC: Modular Accounts with Delegated Validation

### PR DESCRIPTION
**Pull request by nerderlyne**
Discussion Link -> 
https://ethereum-magicians.org/t/erc-7582-modular-accounts-with-delegated-validation/17640

### CODE REVIEW

File: ERCS/erc-7570.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
+---
+eip: 7570
```
Comment (abcoathup) [inline ERCS/erc-7570.md:2]:
```suggestion
eip: 7582
```
Assigning next sequential EIP/ERC/RIP number.  ERC7570 has already been assigned.
Numbering changed to sequential from 7500 and is no longer the PR number.

Please also update the filename.

File: ERCS/erc-7582.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        uint256 actualGasUsed
+    );
+```
+
+[NonceManager](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/core/NonceManager.sol)
+
+```solidity
+    function getNonce(address sender, uint192 key /*validator data*/) external view returns (uint256 nonce);
+```
+
```
Comment (SamWilsn) [inline ERCS/erc-7582.md:85]: You'll need a Rationale section, where you explain any technical decisions you've made within your proposal.

File: ERCS/erc-7582.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+## Backwards Compatibility
+
+No backward compatibility issues found.
+
+## Reference Implementation
+
+See `https://github.com/NaniDAO/7582-account`
```
Comment (SamWilsn) [inline ERCS/erc-7582.md:92]: You'll either need to add your reference implementation into your `assets/` directory, or remove this section entirely:

```suggestion
```

File: ERCS/erc-7582.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        address indexed sender,
+        address indexed paymaster,
+        uint256 nonce, /*validator data*/
+        bool success,
+        uint256 actualGasCost,
+        uint256 actualGasUsed
+    );
+```
+
+[NonceManager](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/core/NonceManager.sol)
```
Comment (SamWilsn) [inline ERCS/erc-7582.md:80]: Please move the interface definition into this proposal, and remove this link. Note that EIPs should only describe the externally visible behaviour of a contract, and not define any implementation details.

File: ERCS/erc-7582.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+```solidity
+    function validateUserOp(
+        UserOperation calldata userOp, /*userOp.nonce: validator data*/
+        bytes32 userOpHash,
+        uint256 missingAccountFunds
+    ) external returns (uint256 validationData);
+```
+
+[`IEntryPoint`](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/interfaces/IEntryPoint.sol):
```
Comment (SamWilsn) [inline ERCS/erc-7582.md:66]: Please refer to the definition in ERC-4337 and not to an external repository.

File: ERCS/erc-7582.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        uint256 verificationGasLimit;
+        uint256 preVerificationGas;
+        uint256 maxFeePerGas;
+        uint256 maxPriorityFeePerGas;
+        bytes paymasterAndData;
+        bytes signature;
+    }
+```
+
+[`IAccount`](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/interfaces/IAccount.sol):
```
Comment (SamWilsn) [inline ERCS/erc-7582.md:56]: Please refer to the definition in ERC-4337 and not to an external repository.

File: ERCS/erc-7582.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+In all of the above validation steps, the validator contract MUST respect the ERC-4337 Entry Point conventions. Note, that while validator key data might be included elsewhere in a `UserOperation` to achieve similar contract account modularity, for example, by packing this data into the `signature` field, this proposal opts to repurpose `nonce` for this pointer to minimize calldata costs and to benefit from the EntryPoint's `getNonce` accounting, as well as the discoverability of user plugin interactions in the `UserOperationEvent` which exposes `nonce` but not other userOp data.
+
+ERC-4337 references:
+
+[UserOperation](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/interfaces/UserOperation.sol):
```
Comment (SamWilsn) [inline ERCS/erc-7582.md:38]: Please refer to the ERC-4337 definition and not to an external repository.

File: ERCS/erc-7582.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+![diagram showing basic 7582 flow](../assets/eip-7582/base-flow.svg)
+
+We leverage the key in ERC-4337 semi-abstracted nonce as the pointer to `validator` address. If a non-sequential key (`>type(uint64).max`) is used as an ERC-4337 Entry Point `UserOperation` (userOp) `nonce`, the `validateUserOp` function in the `sender` contract account MUST extract the validator contract address paired to this key from storage. If the validator contract address is found (i.e., no null return), the ERC7582 contract account MUST forward the userOp calldata to the validator. This calldata SHOULD be the entire userOp. In response to this delegated validation, the validator contract MUST return the ERC-4337 `validationData`, and the ERC7582 `sender` account MUST return this as the `validationData` to the Entry Point.
```
Comment (SamWilsn) [inline ERCS/erc-7582.md:32]: We prefer not self-referencing the current document by number if possible. Instead use "this proposal" or similar.

In this particular case, you could define a term for compliant contracts (maybe "MADV contracts"?) and use that throughout.

**Review by SamWilsn — CHANGES_REQUESTED** [review]
Looking pretty good! Just need to remove external links and address a few minor things.

---

I hope you don't mind, but I took the liberty of fixing a few ERC-4337 references and made your SVG work in light mode as well.

File: ERCS/erc-7582.md (hunk 2 above)
Comment (nerderlyne) [reply ERCS/erc-7582.md:85]: Thank you for the review! I will add a rationale section.

File: ERCS/erc-7582.md (hunk 3 above)
Comment (nerderlyne) [reply ERCS/erc-7582.md:92]: Can the reference implementation in `assets/` directory link to external directories? In this case to 4337 BaseAccount implementation.

File: ERCS/erc-7582.md (hunk 8 above)
Comment (nerderlyne) [reply ERCS/erc-7582.md:32]: Thank you I will keep that in mind going forward. **MADV contracts** is perfect.

File: ERCS/erc-7582.md (hunk 3 above)
Comment (SamWilsn) [reply ERCS/erc-7582.md:92]: We prefer self contained reference implementations. You can do a hand-wavy reference to another project (`import { IERC20 } from '@openzeppelin/contracts/token/ERC20/IERC20.sol';` is pretty common), but avoid any explicit URLs.

File: assets/erc-7582/MADVAccount.sol [hunk 9]
```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: AGPL-3.0-only
```
Comment (SamWilsn) [inline assets/erc-7582/MADVAccount.sol:1]: Unfortunately we cannot use GPL or other copyleft code as reference implementations. Doing so would mean anyone who looks at the reference implementation would also have to release their code as GPL.

Personally, I love the GPL, but it doesn't make sense in a standards repository.

**Review by SamWilsn — CHANGES_REQUESTED** [review]

File: assets/erc-7582/MADVAccount.sol (hunk 9 above)
Comment (nerderlyne) [reply assets/erc-7582/MADVAccount.sol]: I have removed the license identifier 👍

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**